### PR TITLE
[packager] breaking test and fix for combination of "browser" and "react-native" fields

### PR DESCRIPTION
**Pull request by mvayngrib**
...in package.json

### CODE REVIEW

File: packager/react-packager/src/DependencyResolver/Package.js [hunk 1]
```diff
@@ -92,9 +92,22 @@ class Package {
...
+  }
+
+  if (typeof browser === 'string') {
+    browser = {[pkg.main]: browser}
+  }
+
+  // merge with "browser" as default,
+  // "react-native" as override
+  return {...browser,...rn}
 }
```
Comment (skevy) [inline packager/react-packager/src/DependencyResolver/Package.js]: Use semicolons please :)

File: packager/react-packager/src/DependencyResolver/Package.js [hunk 2]
```diff
@@ -92,9 +92,22 @@ class Package {
...
+  let browser = pkg.browser
+  if (rn == null) return browser
+  if (browser == null) return rn
+
+  if (typeof rn === 'string') {
+    rn = {[pkg.main]: rn}
+  }
+
+  if (typeof browser === 'string') {
+    browser = {[pkg.main]: browser}
```
Comment (skevy) [inline packager/react-packager/src/DependencyResolver/Package.js]: nit: space around curly braces

File: packager/react-packager/src/DependencyResolver/Package.js [hunk 3]
```diff
@@ -92,9 +92,22 @@ class Package {
...
-  return pkg['react-native'] == null
-    ? pkg.browser
-    : pkg['react-native'];
+  let rn = pkg['react-native']
+  let browser = pkg.browser
+  if (rn == null) return browser
+  if (browser == null) return rn
+
+  if (typeof rn === 'string') {
+    rn = {[pkg.main]: rn}
```
Comment (skevy) [inline packager/react-packager/src/DependencyResolver/Package.js]: nit: space around braces

File: packager/react-packager/src/DependencyResolver/Package.js [hunk 4]
```diff
@@ -92,9 +92,22 @@ class Package {
 }
 
 function getReplacements(pkg) {
-  return pkg['react-native'] == null
-    ? pkg.browser
-    : pkg['react-native'];
+  let rn = pkg['react-native']
+  let browser = pkg.browser
+  if (rn == null) return browser
```
Comment (skevy) [inline packager/react-packager/src/DependencyResolver/Package.js]: Don't use short version of `if`...use curly braces and put the return on a new line.

**Comment by mvayngrib** [comment]
@skevy fixed style, eslint seems to like it :)

**Comment by mkonicek** [comment]
Hey @mvayngrib, what is a "browser" field? Can you explain what problem this PR solves?

**Comment by mkonicek** [comment]
It looks like you might have to rebase.

**Comment by mvayngrib** [comment]
@mkonicek re: what is the "browser" field - many npm modules don't work in the browser out of the box, e.g. ones depending on node's crypto module. The "browser" field is a way for a module creator to specify what shim/polyfill/etc to use instead of a given file or depedency, if the module is being packaged for a browser-like environment.

The "react-native" field is something I added in another PR, because the browser and react-native environments are significantly different, and different kinds of shims are needed or even possible. For example [udp](https://github.com/tradle/react-native-udp) is possible in react-native but not in the browser.

However, as the "react-native" field is relatively new, the "browser" field is a used as a fallback. This PR fixes a bug in the merging of the primary "react-native" field mapping and the "browser" fallback one. Specifically, when a mapping exists in both fields, the "react-native" one should be used over the "browser" one.

**Comment by facebook-github-bot-6** [comment]
@AaaChiuuu would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by mvayngrib** [comment]
actually, i already moved it here: https://github.com/facebook/node-haste/pull/61

**Comment by mkonicek** [comment]
@mvayngrib OK thanks for the update, closing this one.

File: packager/react-packager/src/DependencyResolver/Package.js [hunk 5]
```diff
@@ -92,9 +92,27 @@ class Package {
 }
 
 function getReplacements(pkg) {
-  return pkg['react-native'] == null
-    ? pkg.browser
-    : pkg['react-native'];
+  let rn = pkg['react-native'];
+  let browser = pkg.browser;
+  if (rn == null) {
+    return browser;
```
Comment (davidaurelio) [inline packager/react-packager/src/DependencyResolver/Package.js]: this would bypass the string check you do in line 109, wouldn’t it?

Comment (mvayngrib) [inline packager/react-packager/src/DependencyResolver/Package.js]: @davidaurelio if rn is null, we don't need to worry about merging rn and browser. But if both are non-null, then line 109 is necessary Also, FYI, i opened the same PR here https://github.com/facebook/node-haste/pull/61 , so we can prob close this one

Comment (davidaurelio) [inline packager/react-packager/src/DependencyResolver/Package.js]: got it

**Pull request closed without merging**